### PR TITLE
refactor(pull_out): consolidate similar function

### PR DESCRIPTION
**Pull request by kyoichi-sugahara**
## Description

In this consolidated function, a boolean parameter named `priority_on_efficient_path` is used to control which loop order to use. If this parameter is true, it will try each planner for all start pose candidates. This corresponds to the behavior of `planWithPriorityOnEfficientPath`. 
On the other hand, if this parameter is false, it will try all start pose candidates for each planner. 
This corresponds to the behavior of `planWithPriorityOnShortBackDistance`.

## Tests performed

Not applicable.

## Effects on system behavior

Not applicable.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp [hunk 1]
```diff
@@ -583,9 +565,9 @@ void PullOutModule::updatePullOutStatus()
   std::vector<Pose> start_pose_candidates = searchPullOutStartPoses();
 
   if (parameters_->search_priority == "efficient_path") {
-    planWithPriorityOnEfficientPath(start_pose_candidates, goal_pose);
+    planWithPriority(start_pose_candidates, goal_pose, true);
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp:568]: I would like to give not bool but `parameters_->search_priority` for adding other policies in the future.

File: planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp [hunk 2]
```diff
@@ -424,97 +425,78 @@ void PullOutModule::planWithPriorityOnEfficientPath(
...
-    if (status_.is_safe) {
-      break;
     }
+    return order_priority;
+  };
+
+  // Choose loop order based on priority_on_efficient_path
+  PriorityOrder order_priority;
+  if (priority_on_efficient_path) {
+    order_priority = makePriorityOrder_PlannerFirst();
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp:493]: I'm not sure but is this function name is ok in terms of naming convention?

File: planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp [hunk 3]
```diff
@@ -424,97 +425,78 @@ void PullOutModule::planWithPriorityOnEfficientPath(
...
-      }
-      // use current path if back is not needed
-      if (status_.back_finished) {
-        status_.is_safe = true;
-        status_.pull_out_path = *pull_out_path;
-        status_.pull_out_start_pose = pull_out_start_pose;
-        status_.planner_type = planner->getPlannerType();
-        break;
-      }
+  auto loop_body = [&](size_t i, const auto & planner) {
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp:428]:
```suggestion
  auto loop_body = [&](const size_t i, const auto & planner) {
```
?

Comment (kyoichi-sugahara) [reply planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp]: :+1:

File: planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp (hunk 2 above)
Comment (kyoichi-sugahara) [reply planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp]: changed to `make_loop_order_planner/pose_first`

File: planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp (hunk 1 above)
Comment (kyoichi-sugahara) [reply planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp]: @kosuke55 
I agree with your opinion.
And could you check my chenge in [the commit](da7e98af06be785e508e274feebbaf4b244668ec) matchs with your image?

Comment (kosuke55) [reply planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp]: thanks, seems good to me!

File: planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp [hunk 4]
```diff
@@ -413,8 +413,9 @@ PathWithLaneId PullOutModule::getCurrentPath() const
   return status_.pull_out_path.partial_paths.at(status_.current_path_idx);
 }
 
-void PullOutModule::planWithPriorityOnEfficientPath(
-  const std::vector<Pose> & start_pose_candidates, const Pose & goal_pose)
+void PullOutModule::planWithPriority(
+  const std::vector<Pose> & start_pose_candidates, const Pose & goal_pose,
+  std::string search_priority)
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp:418]:
```suggestion
  const std::string search_priority)
```

File: planning/behavior_path_planner/include/behavior_path_planner/scene_module/pull_out/pull_out_module.hpp [hunk 5]
```diff
@@ -129,10 +129,9 @@ class PullOutModule : public SceneModuleInterface
 
   void incrementPathIndex();
   PathWithLaneId getCurrentPath() const;
-  void planWithPriorityOnEfficientPath(
-    const std::vector<Pose> & start_pose_candidates, const Pose & goal_pose);
-  void planWithPriorityOnShortBackDistance(
-    const std::vector<Pose> & start_pose_candidates, const Pose & goal_pose);
+  void planWithPriority(
+    const std::vector<Pose> & start_pose_candidates, const Pose & goal_pose,
+    std::string search_priority);
```
Comment (kosuke55) [inline planning/behavior_path_planner/include/behavior_path_planner/scene_module/pull_out/pull_out_module.hpp:134]:
```suggestion
    const std::string search_priority);
```

**Review by kosuke55 — APPROVED** [review]
LGTM, thanks very much!!

**Pull request merged**